### PR TITLE
Google Cloud Platform friendly deployment

### PR DESCRIPTION
**Pull request by saveav**
## Desc

Basically: Make the freegeoip deployable to GCP using Kubernetes.

There are two services offered by GCP that allow us to deploy an app on the platform:

1. Google App Engine. More like Heroku with a lot of 'but...'
2. Google Compute Engine. More like AWS EC2/DigitalOcean Droplet.

The cheapest option is using F1 instance of GCE, but this instance is very unstable to be used to run this service, as [F1 can utilize 20% of CPU](https://www.opsdash.com/blog/google-cloud-f1-micro.html) although it has periodical micro bursting capabilities.

The next option, [G1 is all-in-all around ~44 USD/mo](https://cloud.google.com/products/calculator/#id=7574a748-fba7-4aa7-8579-a4ae26ab97d0) for 3 nodes with 50GB permanent disk attached, which is around the same price of Google App Engine's solution. So, if Google App Engine is like Heroku, where one can just deploy and forget, and it's actually of the same range of price with GCE, then why don't we use it?

Google App Engine is not as _flexible_ as Heroku, so much that they have two categories when we refer to GAE:

1. [Standard environment](https://cloud.google.com/appengine/docs/standard/)
2. [Flexible environment](https://cloud.google.com/appengine/docs/flexible/)

Standard environment is only offered for app written in the following languages:

- Python
- Java
- Go
- PHP

Rumors has it that GAE is Google-first, meaning, services written by and for Google may be written in a standard GAE-compatible way... with their SDK. Yes, to write in that environment, we have to tie our app with [some construct](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/appengine/gophers/gophers-1/main.go) [of their SDK](https://cloud.google.com/appengine/docs/standard/go/quickstart).

In flexible environment, things are much more flexible, but not really. The thing is GAE has no concept of buildpack. So for example, if we would like to deploy Rails, [we still have to do a lot of things manually](https://cloud.google.com/ruby/rails/appengine).

So, either using standard or flexible environment, we will still have to write some glue-code, and possibly changing the code of this project. Which, I will perhaps spend some time trying. So far though, I am really satisfied with the Kubernetes style™️.

Okay, let's forget GAE for now then due to the issue raised above. Luckily, now we are living in the age of containerization. With Kubernetes, we can make those containers deployable to almost any cloud supporting Kubernetes ([AWS supports it](https://aws.amazon.com/eks/)). Kubernetes also make it really easy to scale, and migrate cluster from one [machine type](https://cloud.google.com/compute/pricing) to another machine type.

So, the idea of working with Kubernetes is as follow:

- K8s helps setup the cluster
  - A cluster is a group of pools of nodes. Its presence is almost always permanent, even if we want to upgrade the machine type.
  - Pool of nodes live in a certain region (eg: `asia-northeast1`), and it specifies the machine type of the nodes in that region, for the said cluster. It's possible to have many pool of nodes spread on different regions.
  - A node live in a certain zone (eg: `asia-northeast1-a`, `asia-northeast1-b`, `asia-northeast1-c`), and a zone belongs exclusively to a region. 
  - [A pod](https://coreos.com/kubernetes/docs/latest/pods.html) lives on those nodes, on the same host, for the container deployed replicated `n` times as wanted. Unlike node, a `Pod` is highly, highly ephemeral/they are very mortal (in a very stark contrast to a cluster)
- We build the docker image locally
- We register the docker image to Google Container Registry (perhaps to other place also OK)
- We make use of configurations to tell K8s what to do with our cluster

There are all kind of sorts of configuration on K8s, among which we make use of are:

- [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/): we use it to deploy our container.
- [Service](https://kubernetes.io/docs/concepts/services-networking/service/): we use it to setup the likes of LoadBalancer, and expose the app to the world.
- [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
- [ConfigMap](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/)

So, with this K8s stuff, we don't actually communicate with the Google Compute Engine directly. In fact, we are working with [Google Kubernetes Engine](https://cloud.google.com/kubernetes-engine/). It is nothing more than just an orchestration service. In fact K8s is not related to Google Cloud Platform at all in anyway, although K8s is born in Google. K8s is now donated to [Cloud Native Computing Foundation](https://techcrunch.com/2015/07/21/as-kubernetes-hits-1-0-google-donates-technology-to-newly-formed-cloud-native-computing-foundation-with-ibm-intel-twitter-and-others/) which allow the likes of AWS to benefit greatly of it as well, since K8s is kind of the standard. An alternative to K8s may be Docker Swarm, which is far simpler and a bit less widely adopted so using it may require a lot of manual setups on our GCE--which sounds like not worth it.

In summary, with Kubernetes, maintaining clusters of nodes across different zones in different regions feel like actually as easy as deploying to heroku! There's no custom glue code, just container, and deployment scripts.

However, we have to setup our machine to recognize K8s as is used with GCP. So, we will have to [install gcloud CLI first](https://cloud.google.com/sdk/docs/quickstart-macos) and then do configuration and init and [install the K8s component with it](https://cloud.google.com/kubernetes-engine/docs/tutorials/hello-app).

When I said 

> K8s helps setup the cluster

I am sorry that I don't actually mean it exclusively. K8s do indeed can help setup the cluster. But... K8s can also _recognize_ the cluster, ONCE it's setup properly with the `gcloud`. So, we can technically setup the cluster through a [beautiful user interface](https://console.cloud.google.com/kubernetes/list) rather than through the command line.

Even though K8s is awesome, in my opinion, it yet to beat Heroku's simplicity. Like, for example, we have to setup logging manually. Google has Stackdriver which is a log-viewer or log aggregator in quite the same sense like Papertrail. To direct the logging from our app to Stackdriver, we need a logging agent and [fluentd](https://cloud.google.com/logging/docs/agent/) is one such. I(r|c)onic, it's written in Ruby. Ruby wins!

## Effect

Can deploy to GCP

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,257 +1,83 @@
...
-```bash
-docker run --restart=always -p 8080:8080 -d apilayer/freegeoip
 ```
-
-#### Test
-
-```bash
-curl localhost:8080/json/1.2.3.4
-# => {"ip":"1.2.3.4","country_code":"US","country_name":"United States", # ...
+$ docker build -t gcr.io/{project-id}/vgeoip:1.0.0 .
```
Comment (ignatiusreza) [inline README.md]: let's actually write down the `{project-id}`, which if i'm not mistaken is `voyagin-prod`? so we can copy paste

**Comment by saveav** [comment]
Image assets:

logs

**Comment by saveav** [comment]
ready for review @ignatiusreza

**Comment by saveav** [comment]
Test comparison between GCP and our PS:

[comparison.zip](https://github.com/voyagin/freegeoip/files/2295978/comparison.zip)

**Comment by ignatiusreza** [comment]
Thanks for the write up @saveav!

I'll add that fluentd itself is not a logging agent.. it's a data collector with extensive and extensible input/output plugin system.. one of the most common usage is to process stream of logs, but its capabilities is not limited to just logging..

one more question, did you hand wrote all of those configuration? if some of it is generated, it would be better to split the commit like:

- `run some generator command`
- `edit some config because of reason`..

that way, we can clearly see, what the default were, and why we need to change it.. assuming that most configuration should have good sane defaults..

**Comment by saveav** [comment]
Yes. actually you are very right, @ignatiusreza . It's a data collector, and it's extensible. But if not mistaken GCP call it as logging agent, not sure what that means...

Some are hand written (especially about deployments). But the one about fluentd, I copy from a repo owned by GCP...

**Comment by saveav** [comment]
Found the original repo: [here](https://github.com/GoogleCloudPlatform/container-engine-customize-fluentd/tree/master/kubernetes).

**Comment by ignatiusreza** [comment]
sample commit message:

```
add fluentd configuration to enable logging in GCE

config files are copied verbatim from

https://github.com/GoogleCloudPlatform/container-engine-customize-fluentd/tree/6a46d72b29f3d8e8e495713bc3382ce28caf744e/kubernetes
```

notice how the link to repo are also included, so that the next person who need to update the config can find it easily to understand where it came from..

**Comment by saveav** [comment]
got it. will do.

yes copied verbatim.

**Comment by saveav** [comment]
since it's copied, so I guess... it's ready for review...

**Comment by saveav** [comment]
I will ask help to make it private.

And let's re-associate it with a new static IP address...

**Comment by letrungkien** [comment]
I think the test GCP vs PS is not fair in term of our setup.
The geoip server is only called by our `prod-voyagin` app so we should open a rails console in Private Space and test the result.

I ran a quick check with this script
```ruby
def lookup(ip)
  start = Time.current
  url = "http://ps-vgeoip.herokuapp.com/json/#{ip}"
  uri = URI(url)
  Net::HTTP.get(uri)
  ((Time.current - start) * 1000).to_i
end

def lookup2(ip)
  start = Time.current
  url = "http://35.200.118.6:8080/json/#{ip}"
  uri = URI(url)
  Net::HTTP.get(uri)
  ((Time.current - start) * 1000).to_i
end

result1 = 100.times.map { |_| lookup(IPAddr.new(rand(2**32),Socket::AF_INET).to_s) }
result2 = 100.times.map { |_| lookup2(IPAddr.new(rand(2**32),Socket::AF_INET).to_s) }
```

`result1` is result of PS geoip
`result2` is result of GCP geoip

with 3 run ( result in ms)

PS | Sum | Max | Min | Avg
--- | --- | --- | --- | ---
Run 1 | 938 | 83 | 3 | 9.38
Run 2 | 857 | 20 | 3 | 8.57
Run 3 | 890 | 44 | 3 | 8.9


GCP | Sum | Max | Min
--- | --- | --- | --- |
Run 1 | 1215 | 18 | 10 | 12.15
Run 2 | 1204 | 16 | 10 | 12.04
Run 3 | 1303 | 104 | 10 | 13.03

Switching from PS geoip to GCP geoip will likely increase about 3ms to 4 ms in average

**Comment by saveav** [comment]
Yes, that's right. I should have tested it in PS too, since its only users are PS-hosted. Thanks for the complimentary tests @letrungkien (this discussion is public 😭) and you released our ps path 😭 I will make a quick edit to censor it lol

thanks anyway. it makes sense. PS-to-PS will be faster. I think, we should be able to tolerate 3ms, but let's not throw away our PS server yet in case there's a misbehavior. 👍

**Comment by saveav** [comment]
github displayed the edit history...

**Comment by letrungkien** [comment]
@saveav Thanks. 😨 

I can delete it and respost here:

I think the test GCP vs PS is not fair in term of our setup.
The geoip server is only called by our `prod-voyagin` app so we should open a rails console in Private Space and test the result.

I ran a quick check with this script
```ruby
def lookup(ip)
  start = Time.current
  url = "http://someserver.herokuapp.com/json/#{ip}"
  uri = URI(url)
  Net::HTTP.get(uri)
  ((Time.current - start) * 1000).to_i
end

def lookup2(ip)
  start = Time.current
  url = "http://35.200.118.6:8080/json/#{ip}"
  uri = URI(url)
  Net::HTTP.get(uri)
  ((Time.current - start) * 1000).to_i
end

result1 = 100.times.map { |_| lookup(IPAddr.new(rand(2**32),Socket::AF_INET).to_s) }
result2 = 100.times.map { |_| lookup2(IPAddr.new(rand(2**32),Socket::AF_INET).to_s) }
```

`result1` is result of PS geoip
`result2` is result of GCP geoip

with 3 run ( result in ms)

PS | Sum | Max | Min | Avg
--- | --- | --- | --- | ---
Run 1 | 938 | 83 | 3 | 9.38
Run 2 | 857 | 20 | 3 | 8.57
Run 3 | 890 | 44 | 3 | 8.9


GCP | Sum | Max | Min | Avg
--- | --- | --- | --- | ---
Run 1 | 1215 | 18 | 10 | 12.15
Run 2 | 1204 | 16 | 10 | 12.04
Run 3 | 1303 | 104 | 10 | 13.03

Switching from PS geoip to GCP geoip will likely increase about 3ms to 4 ms in average

**Comment by saveav** [comment]
updated the readme to include how the service works...

**Comment by saveav** [comment]
provider | median | average | max | min
-------- | -------- | -------- | ---- | ----- |
zeit | 480ms | 484.13ms | 651ms | 471ms
priv. space | 8ms | 8.2ms | 16ms | 8ms

**Comment by saveav** [comment]
I am closing this since we are beginning to use AWS Fargate

**Pull request closed without merging**